### PR TITLE
cancellation file support for bluebox

### PR DESCRIPTION
**Pull request by wjblanke**
This change allows a file path to be passed in which must exist for a GMP slow blue box VDF to continue running. This enables the calling process to abort the VDF by removing the file (a temporary file so crashes or bad cleanups do not affect future runs). This is important to allow bluebox timelord processes to exit cleanly and not require termination.

Associated chia blockchain pr (without VDF update so broken)
https://github.com/Chia-Network/chia-blockchain/pull/18642

### CODE REVIEW

**Review by arvidn — DISMISSED** [review]

File: src/c_bindings/c_wrapper.cpp [hunk 1]
```diff
@@ -22,7 +22,7 @@ extern "C" {
             std::vector<uint8_t> challenge_hash_bytes(challenge_hash, challenge_hash + challenge_size);
             integer discriminant = CreateDiscriminant(challenge_hash_bytes, discriminant_size_bits);
             form x = DeserializeForm(discriminant, x_s, x_s_size);
-            std::vector<uint8_t> result = ProveSlow(discriminant, x, num_iterations);
+            std::vector<uint8_t> result = ProveSlow(discriminant, x, num_iterations, "");
```
Comment (arvidn) [inline src/c_bindings/c_wrapper.cpp:25]: ideally, `prove_wrapper()` would take a `char const*` and you would turn that into a `std::string` (where `nullptr` means an empty string). We're working towards using the rust binding, so we'd have to add this eventually anyway

File: src/prover_slow.h [hunk 2]
```diff
@@ -100,6 +101,19 @@ std::vector<uint8_t> ProveSlow(integer& D, form& x, uint64_t num_iterations) {
...
         nudupl_form(y, y, D, L);
         reducer.reduce(y);
+
+        // Check for cancellation every 65535 interations
+        if ((i&0xffff)==0) {
+            // Only if we have a shutdown path
+            if (shutdown_file_path!="") {
+                struct stat buffer;
+            
+                if (stat (shutdown_file_path.c_str(), &buffer) != 0) {
```
Comment (arvidn) [inline src/prover_slow.h:111]: there's a slight risk that this call fails with `EINTR`, which probably should not be interpreted as a request to abort.

File: src/c_bindings/c_wrapper.cpp (hunk 1 above)
Comment (wjblanke) [reply src/c_bindings/c_wrapper.cpp:25]: unfortunately i couldn't get the rust tests to work due to my lack of rust skills, thats why it is hacked to the original implementation here

File: src/prover_slow.h (hunk 2 above)
Comment (wjblanke) [reply src/prover_slow.h]: fixing this thanks!

File: src/prover_slow.h [hunk 3]
```diff
@@ -108,7 +108,8 @@ std::vector<uint8_t> ProveSlow(integer& D, form& x, uint64_t num_iterations, std
             if (shutdown_file_path!="") {
                 struct stat buffer;
             
-                if (stat (shutdown_file_path.c_str(), &buffer) != 0) {
+                int statrst = stat(shutdown_file_path.c_str(), &buffer);
+                if ((statrst != 0) && (statrst != EINTR)) {
```
Comment (arvidn) [inline src/prover_slow.h:112]:
```suggestion
                if ((statrst != 0) && (errno != EINTR)) {
```

Comment (arvidn) [reply src/prover_slow.h:112]: https://man7.org/linux/man-pages/man2/stat.2.html

Comment (wjblanke) [reply src/prover_slow.h]: oops!

**Review by arvidn — APPROVED** [review]

**Review by arvidn — APPROVED** [review]

**Pull request merged**